pkg/resourcemanager/rediscaches/rediscacheaction_reconcile.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package rediscaches
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-service-operator/api/v1alpha1"
+	"github.com/Azure/azure-service-operator/pkg/resourcemanager"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func (m *AzureRedisCacheActionManager) Ensure(ctx context.Context, obj runtime.Object, opts ...resourcemanager.ConfigOption) (bool, error) {
+	options := &resourcemanager.Options{}
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	instance, err := m.convert(obj)
+	if err != nil {
+		return true, err
+	}
+
+	// never re-provision an action
+	if instance.Status.Provisioned {
+		return true, nil
+	}
+
+	rollAllKeys := instance.Spec.ActionName == v1alpha1.RedisCacheActionNameRollAllKeys
+
+	if rollAllKeys || instance.Spec.ActionName == v1alpha1.RedisCacheActionNameRollPrimaryKey {
+		if err = m.RegeneratePrimaryAccessKey(ctx, instance.Spec.ResourceGroup, instance.Spec.CacheName); err != nil {
+			return false, err
+		}
+	}
+
+	if rollAllKeys || instance.Spec.ActionName == v1alpha1.RedisCacheActionNameRollSecondaryKey {
+		if err = m.RegenerateSecondaryAccessKey(ctx, instance.Spec.ResourceGroup, instance.Spec.CacheName); err != nil {
+			return false, err
+		}
+	}
```

Comment:
NIT Alert: This might be one of those rare occasions where a switch is actually cleaner...let's see...

```
switch instance.Spec.ActionName {
case v1alpha1.RedisCacheActionNameRollAllKeys, v1alpha1.RedisCacheActionNameRollPrimaryKey:
              if err = m.RegeneratePrimaryAccessKey(ctx, instance.Spec.ResourceGroup, instance.Spec.CacheName); err != nil {
			return false, err
		}
case v1alpha1.RedisCacheActionNameRollAllKeys, v1alpha1.RedisCacheActionNameRollSecondaryKey:
               if err = m.RegenerateSecondaryAccessKey(ctx, instance.Spec.ResourceGroup, instance.Spec.CacheName); err != nil {
			return false, err
		}
}
```

Reply:
Wouldn't that only regenerate the primary key when you try to "rollallkeys" because the first case succeeds and doesn't hit the second case?

Reply:
ah, could be, probably easy enough to test